File: Python/DoublyLinkedList.py
class Node:
    """ A node in doubly linked list """
    def __init__(self, data, prev=None, next=None):
        self.prev = prev
        self.data = data
        self.next = next

    def __repr__(self):
        return repr(self.data)


class DoublyLinkedList:
    def __init__(self):
        """ 
        Create a new doubly linked list
        Time O(1)
        """
        self.head = None

    def __repr__(self):
        """ 
        Return a string representation of the list
        Time O(n)
        """
        nodes = []
        curr = self.head
        while curr:
            nodes.append(repr(curr))
            curr = curr.next

        return '[ ' + ', '.join(nodes) + ' ]'

    def pushFront(self, data):
        """
        Insert a new element at the beginning
        Time O(1)
        """
        if self.head is None:
            node = Node(data, None, None)
        else:
            node = Node(data, None, self.head)
            self.head.prev = node
        self.head = node

    def topFront(self):
        """
        Return the first element of the list
        Time O(1)
        """
        return self.head

    def popFront(self):
        """ 
        Remove the first element of the list
        Time O(1)
        """
        self.head.next.prev = None
        self.head = self.head.next

    def pushBack(self, data):
        """
        Add a new element at the end of the list
        Time O(n)
        """
        if not self.head:
            self.head = Node(data)
            return
        
        curr = self.head
        while curr.next:
            curr = curr.next

        curr.next = Node(data, curr)

    def topBack(self):
        """
        Returns the last element of the list
        Time O(n)
        """
        curr = self.head
        while curr.next:
            curr = curr.next
        return curr.data

    def popBack(self):
        """
        Removes the last element of the list
        Time O(n)
        """
        curr = self.head
        while curr.next.next:
            curr = curr.next
        curr.next = None

    def isEmpty(self):
        """
        Check if list is empty or not
        Time O(1)
        """
        return True if self.head is None else False

    def addBefore(self, key, data):
        """
        Add an element before given key
        Time O(n)
        """
        curr = self.head
        if key == curr.data:
            self.pushFront(data)
            return
        while curr:
            if curr.next is None:
                break
            if curr.next.data == key:
                node = Node(data, curr.next.prev, curr.next)
                curr.next = node
                break
            else:
                curr = curr.next
        return

    def addAfter(self, key, data):
        """
        Add a new element after given key
        Time O(n)
        """
        curr = self.head
        while curr:
            if curr.data == key:
                node = Node(data, curr, curr.next)
                curr.next = node
                break
            else:
                curr = curr.next
        return

    def find(self, key):
        """
        Search for the particular key
        Time O(n)
        """
        curr = self.head
        while curr and curr.data != key:
            curr = curr.next

        return True if curr else False

    def remove(self, key):
        """
        Removes the first occurrence of the key
        Time O(n)
        """
        curr = self.head
        while curr and curr.data != key:
            curr = curr.next
        
        if not curr:
            return
        
        if curr.prev is None:
            self.head = curr.next
        elif curr.next is None:
            curr.prev.next = curr.next
        else:
            curr.prev.next = curr.next
            curr.next.prev = curr.prev
        return


if __name__ == '__main__':
    dlist = DoublyLinkedList()
    dlist.pushFront(10)
    dlist.pushFront(20)
    dlist.pushFront(30)
    dlist.pushFront(40)
    print(dlist)
    dlist.remove(10)
    print(dlist)

File: c++/DoublyLinkedList.cpp
#include<iostream>
using namespace std;

class Node{
public:
    Node* prev;
    int data;
    Node* next;
};

class DoublyLinkedList{
public:
    int length;
    Node* head;

    DoublyLinkedList();
    ~DoublyLinkedList();

    void pushFront(int data);
    int topFront();
    void popFront();
    void pushBack(int data);
    int topBack();
    void popBack();
    void find(int key);
    void erase(int key);
    bool isEmpty();
    void addBefore(int key, int data);
    void addAfter(int key, int data);
    void print();
};

DoublyLinkedList::DoublyLinkedList(){
    this->length = 0;
    this->head = NULL;
}

DoublyLinkedList::~DoublyLinkedList(){
    cout << "List Deleted" << endl;
}

void DoublyLinkedList::pushFront(int data){
    Node* node = new Node();
    node->data = data;
    node->prev = NULL;

    if(this->head){
        node->next = this->head;
        this->head->prev = node;
    }else{
        node->next = NULL;
    }
    
    this->head = node;
    this->length++;
}

int DoublyLinkedList::topFront(){
    return this->head->data;
}

void DoublyLinkedList::popFront(){
    this->head->next->prev = NULL;
    this->head = this->head->next;
}

void DoublyLinkedList::pushBack(int data){
    Node *current = this->head;
    while(current->next){
        current = current->next;
    }
    Node* node = new Node();
    node->data = data;
    node->prev = current;
    current->next = node;
}

int DoublyLinkedList::topBack(){
    Node *current = this->head;

    if(!current){
        return -1;
    }
    while(current->next){
        current = current->next;
    }
    return current->data;
}

void DoublyLinkedList::popBack(){
    Node *current = this->head;
    while (current->next->next)
    {
        current = current->next;
    }
    current->next = NULL;
}

void DoublyLinkedList::find(int key){
    Node *current = this->head;
    while(current != NULL && current->data != key){
        current = current->next;
    }
    if(current){
        printf("Found Node: %d\n", current->data);
    }else{
        printf("Not Found\n");
    }
}

void DoublyLinkedList::erase(int key){
    Node *current = this->head;
    while(current != NULL && current->data != key){
        current = current->next;
    }
    if(current == head){
        current->next->prev = NULL;
        this->head = current->next;
        return;
    }
    if(current->next == NULL){
        current->prev->next = current->next;
        return;
    }
    current->prev->next = current->next;
    current->next->prev = current->prev;    
    return;
}

bool DoublyLinkedList::isEmpty(){
    return this->head == NULL;
}

void DoublyLinkedList::addBefore(int key, int data){
    Node *current = this->head;
    while(current != NULL && current->data != key){
        current = current->next;
    }
    Node *node = new Node();
    node->data = data;

    if(current == head){
        this->head = node;
    }else{
        current->prev->next = node;
    }
    node->prev = current->prev;
    node->next = current;
    current->prev = node;
}

void DoublyLinkedList::addAfter(int key, int data){
    Node *current = this->head;
    while(current != NULL && current->data != key){
        current = current->next;
    }
    Node *node = new Node();
    node->data = data;
    node->prev = current;
    node->next = current->next;

    if(current->next != NULL){
        current->next->prev = node;
    }
    current->next = node;
}

void DoublyLinkedList::print(){
    Node* head = this->head;
    while(head){
        cout << head->data << " => ";
        head = head->next;
    }
    cout << "NULL" << endl;
}

int main()
{
    DoublyLinkedList* dlist = new DoublyLinkedList();
    for(int i = 0; i < 5; ++i){
        dlist->pushFront(i);
    }
    dlist->print();
    printf("Length : %d\n", dlist->length);
    printf("TopFront : %d\n", dlist->topFront());
    printf("TopBack : %d\n", dlist->topBack());
    printf("Is Empty: %d\n", dlist->isEmpty());
    // dlist->popFront();
    // dlist->popBack();
    dlist->pushBack(10);
    dlist->pushBack(20);

    dlist->find(22);
    dlist->erase(20);

    dlist->addBefore(4, 44);

    dlist->print();
    delete dlist;
    return 0;
}